src/tariff_region.cc

```diff
@@ -0,0 +1,164 @@
+#include "tariff_region.h"
+
+namespace cycamore {
+
+TariffRegion::TariffRegion(cyclus::Context* ctx)
+: cyclus::Region(ctx) {}
+
+TariffRegion::~TariffRegion() {}
+
+void TariffRegion::EnterNotify() {
+  Region::EnterNotify();
+  BuildRegionSet();
+  ValidateConfiguration();
+  BuildTariffRules();
+}
+
+void TariffRegion::Tock() {
+    RecordTariffConfiguration();
+}
+
+// Actual implementation of the DRE Functions using the template function:
+void TariffRegion::AdjustMatlPrefs(cyclus::PrefMap<cyclus::Material>::type& prefs) {
+  AdjustPrefsImpl<cyclus::Material>(prefs);
+}
+
+void TariffRegion::AdjustProductPrefs(cyclus::PrefMap<cyclus::Product>::type& prefs) {
+  AdjustPrefsImpl<cyclus::Product>(prefs);
+}
+
+void TariffRegion::BuildRegionSet() {
+  // Get all agents in the simulation
+  const std::set<cyclus::Agent*>& all_agents = context()->GetAgentList();
+  
+  // Find all regions that match our region_names list
+  for (const auto& agent : all_agents) {
+    if (agent->kind() == "Region") {
+      cyclus::Region* region = static_cast<cyclus::Region*>(agent);
+      // Check if this region's prototype matches any in our list
+      auto it = std::find(region_names.begin(), region_names.end(), 
+                         region->prototype());
+      if (it != region_names.end()) {
+        adjustment_regions_.insert(region);
+      }
+    }
+  }
+}
+
+cyclus::Region* TariffRegion::FindMatchingRegion(cyclus::Facility* supplier) {
+  // Get all parent regions of the supplier
+  std::vector<cyclus::Region*> parent_regions = supplier->GetAllParentRegions();
+  
+  // Check each parent region to see if it's in our tariff list
+  for (cyclus::Region* parent_region : parent_regions) {
+    if (adjustment_regions_.find(parent_region) != adjustment_regions_.end()) {
+      return parent_region;
+    }
+  }
+  
+  // No matching region found
+  return nullptr;
+}
+
+void TariffRegion::BuildTariffRules() {
+  tariff_rules_.clear();
+  region_flat_adjustments_map_.clear();
+  
+  // Build tariff rules from the input configuration
+  for (size_t i = 0; i < region_names.size(); ++i) {
+    std::string region_name = region_names[i];
+    double flat_adjustment = (i < region_flat_adjustments.size()) ? 
+                             region_flat_adjustments[i] : 0.0;
+    
+    // Store flat adjustment for this region
+    region_flat_adjustments_map_[region_name] = flat_adjustment;
+    
+    // Figure out where in the list of commodities we start for this region
+    // Necessary because of the way we've implemented the input file
+    if (i < region_commodity_counts.size()) {
+      size_t start_index = 0;
+      for (size_t j = 0; j < i; ++j) {
+        if (j < region_commodity_counts.size()) {
+          start_index += region_commodity_counts[j];
+        }
+      }
```

Comment:
You should be able to initialize this before you start the top level loop and update it appropriately as you march through the vector.